Network Router: Use API; Fix fixed IPs subnet

gildub opened the pull request:
Networks > Network Routers

Replaces http calls in RoR controller by API calls 
See https://github.com/ManageIQ/manageiq-ui-classic/issues/737
Also:
* Filtering only external network and subsequently better proper subnet (floating IPs) handling.
* External gateway displays only when external management provider id has been selected
* Changes source NAT field position as it depends on the network id
* Adds administrative state field

Before:
![router1](https://user-images.githubusercontent.com/1901741/33698753-a3e6b854-db62-11e7-9055-b0337a48d1d6.png)

Now:
![network-routers-after](https://user-images.githubusercontent.com/1901741/39090428-1d93230c-4622-11e8-9e96-76b13c300bb3.png)

gildub commented:
@mansam, could you please assign/review? Thanks

mansam commented:
Looks like maybe the routes for `network_router_form_fields` and other removed controller methods should be removed as part of this PR.

gildub commented:
@mansam, right, was on my list but forgot. Thanks

gildub commented:
@mansam, I removed the routes, thanks.

gildub commented:
More PRs to follow for the controller to component' and then move the rest of the http requests (create/update/delete) to API only.

@martinpovolny, @himdel, @ZitaNemeckova

gildub commented:
@miq-bot add_label refactoring

gildub commented:
@mansam, @himdel, @martinpovolny, could you please assign someone for review please. Thanks

martinpovolny commented:
@AparnaKarve, @Hyperkid123, @ZitaNemeckova  : please, review

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,33 +1,70 @@
-      .catch(miqService.handleFailure);
+    API.get("/api/network_routers/" + networkRouterFormId + "?attributes=name,ems_id,admin_state_up,cloud_network_id,extra_attributes,cloud_tenant,ext_management_system,cloud_subnets").then(function(data) {
+      Object.assign(vm.networkRouterModel, data);
+      if (data.extra_attributes.external_gateway_info && data.extra_attributes.external_gateway_info !== {}) {
```
Be careful `{} === {}` is `false` . You can use lodash function `_.isEmpty(data.extra_attributes.external_gateway_info)` to find out if it is an empty hash.

gildub replied:
Ok, using lodash, thanks

ZitaNemeckova commented:
Before:
<img width="1190" alt="screen shot 2017-11-06 at 9 36 43 am" src="https://user-images.githubusercontent.com/9210860/32432068-5cfeef78-c2d6-11e7-8d4c-046c47ec2ac7.png">
After:
<img width="1199" alt="screen shot 2017-11-06 at 9 38 53 am" src="https://user-images.githubusercontent.com/9210860/32432069-5fe7e2c6-c2d6-11e7-9384-d83c986ddd8a.png">

Buttons `Save` and `Reset` shouldn't be enabled if no change was made.

Code wise:

Looking for values at `angularForm` is not good. Please save those values in controller so they can be accessed via `vm`.

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,33 +1,70 @@
+
+  var getCloudSubnetsByNetworkID = function(id) {
+    if (id) {
+      API.get("/api/cloud_subnets?expand=resources&attributes=name,ems_ref&filter[]=cloud_network_id=" + id).then(function(data) {
```
If you want to wait for API call to be finished before you leave this function it should be `return API.get(...`

gildub replied:
Using better vm.modelCopy after rebase. 

I've started using init() in other places, done it here too.

> Looking for values at angularForm is not good.
Could you please indicate where that's happening?

himdel replied:
Missing `return`

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,33 +1,70 @@
 
+  var getCloudNetworksByEms = function(id) {
+    if (id) {
+      API.get("/api/cloud_networks?expand=resources&attributes=name,ems_ref&filter[]=external_facing=true&filter[]=ems_id=" + id).then(function(data) {
```
If you want to wait for API call to be finished before you leave this function it should be `return API.get(...`

gildub replied:
Using return with `Promise` allows to pass the value to the next `then`, there is not need here for passing such data which only populate fields (or select lists).

ZitaNemeckova replied:
I wasn't clear, sorry :) It's not about the data being sent back but about having control over them and knowing that you already have them. You want to populate fields before you display form and switch off spinner. Without suggested `return`s there is no guarantee that all fields will be populated before user sees them and they may change under his nose. Please add `return`s :)

gildub replied:
I see what you mean with the chaining happening only if Promise is returned. Thank you!

himdel replied:
Misssing `return`

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,33 +1,70 @@
+    }
+  };
+
+  var getSubnetByRef = function(ref) {
```
Same here.

gildub replied:
Same as above too :)

himdel replied:
Missing `return`

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,33 +1,70 @@
+      }
+    }).then(function() {
+      if (vm.networkRouterModel.external_gateway) {
+        getSubnetByRef(vm.networkRouterModel.extra_attributes.external_gateway_info.external_fixed_ips[0].subnet_id);
```
`return getSubnetByRef(...` so it waits on `Promise` that you should return from this function. If you want one `.then` to be finished before next one starts.

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,33 +1,70 @@
+  };
+
   vm.addClicked = function() {
     var url = 'create/new' + '?button=add';
```
This should be one string. No need for `+`. Could you please fix it? :)

gildub replied:
Done.

gildub commented:
@ZitaNemeckova, please review all suggestions have been addressed. I can't wait to have migrated all those networks controller to components, especially from a test view point.

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -65,47 +104,24 @@ ManageIQ.angular.app.controller('networkRouterFormController', ['$http', '$scope
+        vm.available_tenants = data.resources;
+      })(id);
+    }
+    miqService.sparkleOff();
```
Also here is no use of `Promise` and no guarantee that all data will be set before `miqService.sparkleOff()`.

ZitaNemeckova replied:
``` js
var providerTenants = miqService.getProviderTenants(function(data) {
  vm.available_tenants = data.resources;
})(id);
Promise.all([getCloudNetworksByEms(id), providerTenants])
  .then(miqService.sparkleOff)
```
But both must return API calls.

gildub replied:
Yes the sparkleOn/Off flow was really bad, and this is actually quite messy in general as. Since API request are already Promise I have simplified it.

himdel replied:
This sparkleOn/sparkleOff pair is still not doing anything and can be removed :)

gildub replied:
Done

ZitaNemeckova commented:
@gildub  I cannot add new network Router.
```
F, [2017-11-17T14:01:19.193444 #44722] FATAL -- : Error caught: [ActiveRecord::RecordNotFound] Couldn't find ExtManagementSystem without an ID
/usr/local/lib/ruby/gems/2.3.0/gems/activerecord-5.0.6/lib/active_record/relation/finder_methods.rb:456:in `find_with_ids'
/usr/local/lib/ruby/gems/2.3.0/gems/activerecord-5.0.6/lib/active_record/relation/finder_methods.rb:66:in `find'
/usr/local/lib/ruby/gems/2.3.0/gems/activerecord-5.0.6/lib/active_record/querying.rb:3:in `find'
/usr/local/lib/ruby/gems/2.3.0/gems/activerecord-5.0.6/lib/active_record/core.rb:151:in `find'
/Users/zita/Desktop/ManageIQ/manageiq-ui-classic/app/controllers/network_router_controller.rb:68:in `create'
```
I see only difference in send data `ems_id: 10000014` => `ext_management_system[]: 10000014`. Is this on purpose?

ZitaNemeckova commented:
@gildub There are still `return`s of Promises missing.

gildub commented:
@ZitaNemeckova, 

* Added missing returns
* Fixed id issue blocking creation
* Placement is showing only after ems_id is provided since it depends upon it
* Renamed getSubnetByRef -> getCloudSubnetsByRef
* Fixed Id when Network changes

Retested create/update

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,35 +1,71 @@
-  ManageIQ.angular.scope = vm;
+  var getCloudNetworksByEms = function(id) {
+    if (id) {
+      API.get("/api/cloud_networks?expand=resources&attributes=name,ems_ref&filter[]=external_facing=true&filter[]=ems_id=" + id).then(function(data) {
```
missing `return`

gildub replied:
Makes sense, thanks

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,35 +1,71 @@
-    miqService.sparkleOn();
+  var getCloudSubnetsByNetworkID = function(id) {
+    if (id) {
+      API.get("/api/cloud_subnets?expand=resources&attributes=name,ems_ref&filter[]=cloud_network_id=" + id).then(function(data) {
```
missing `return`

gildub replied:
Done

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,35 +1,71 @@
-  }
+  var getCloudSubnetsByRef = function(ref) {
+    if (ref) {
+      API.get("/api/cloud_subnets?expand=resources&attributes=name&filter[]=ems_ref=" + ref).then(function(data) {
```
missing return

gildub replied:
Done

gildub commented:
@ZitaNemeckova, should be good to go, thanks.

himdel commented:
:+1: List of managers is filtered by openstack.
:+1: `ems_id` changes to `ext_management_system[id]` - in js *and* in ruby
? we now also send extra `cloud_tenant:` and `extra_attributes:` (both empty)  (the code looks like it expected `cloud_tenant_id` at some point?)

himdel commented:
.. But I think something is still broken while editing:

Editing the same network router, just changing the name.

Before this PR:

```
name:cloud-routerxx
cloud_subnet_id:10000000000056
ems_id:10000000000014
cloud_network_id:10000000000052
external_gateway:true
enable_snat:true
```

After:

```
name:cloud-routerxx
cloud_subnet_id:10000000000056
cloud_tenant[href]:http://localhost:3000/api/cloud_tenant/10000000000004
cloud_tenant[id]:10000000000004
cloud_tenant[name]:cloud-southeast
cloud_tenant[description]:
cloud_tenant[enabled]:true
cloud_tenant[ems_ref]:7288d564106f4046b3898f3c642cf387
cloud_tenant[ems_id]:10000000000013
cloud_tenant[created_at]:2017-05-31T22:33:02Z
cloud_tenant[updated_at]:2017-06-29T16:12:23Z
cloud_tenant[type]:ManageIQ::Providers::Openstack::CloudManager::CloudTenant
enable_snat:true
external_gateway:true
extra_attributes[distributed]:false
extra_attributes[high_availability]:false
extra_attributes[external_gateway_info][network_id]:6b3d0c3b-8b68-4c26-a493-5be44d160241
extra_attributes[external_gateway_info][enable_snat]:true
extra_attributes[external_gateway_info][external_fixed_ips][0][subnet_id]:1ca5cc3e-ffe1-44cf-94df-98a798489d06
extra_attributes[external_gateway_info][external_fixed_ips][0][ip_address]:10.9.60.154
href:http://localhost:3000/api/network_routers/10000000000006
ems_id:10000000000014
admin_state_up:t
cloud_network_id:10000000000052
id:10000000000006
ext_management_system[href]:http://localhost:3000/api/ext_management_system/10000000000014
ext_management_system[id]:10000000000014
ext_management_system[name]:OpenStack Network Manager
ext_management_system[created_on]:2017-06-01T04:33:30Z
ext_management_system[updated_on]:2017-07-12T20:12:11Z
ext_management_system[zone_id]:10000000000005
ext_management_system[type]:ManageIQ::Providers::Openstack::NetworkManager
ext_management_system[provider_region]:
ext_management_system[last_refresh_date]:2017-07-12T20:12:11Z
ext_management_system[tenant_id]:10000000000001
ext_management_system[parent_ems_id]:10000000000013
ext_management_system[tenant_mapping_enabled]:true
ext_management_system[enabled]:true
cloud_subnets[0][href]:http://localhost:3000/api/cloud_subnets/10000000000059
cloud_subnets[0][id]:10000000000059
cloud_subnets[0][name]:cloud-subnet
cloud_subnets[0][ems_ref]:e3eaa57d-8cfd-456a-944a-a28c65784a8e
cloud_subnets[0][ems_id]:10000000000014
cloud_subnets[0][cloud_network_id]:10000000000055
cloud_subnets[0][cidr]:192.168.76.0/24
cloud_subnets[0][status]:active
cloud_subnets[0][dhcp_enabled]:true
cloud_subnets[0][gateway]:192.168.76.1
cloud_subnets[0][network_protocol]:ipv4
cloud_subnets[0][cloud_tenant_id]:10000000000004
cloud_subnets[0][dns_nameservers][]:10.11.164.2
cloud_subnets[0][extra_attributes][allocation_pools][0][start]:192.168.76.2
cloud_subnets[0][extra_attributes][allocation_pools][0][end]:192.168.76.254
cloud_subnets[0][extra_attributes][ip_version]:4
cloud_subnets[0][type]:ManageIQ::Providers::Openstack::NetworkManager::CloudSubnet
cloud_subnets[0][network_router_id]:10000000000006
```

Clearly you are sending more data than you should :).

gildub commented:
@himdel, well that's the trade off of dealing with objects and their dependencies especially when using `Object.assign`. Okay let me filter down the data starting with the API fetch.

himdel commented:
@gildub Yeah, I know.. I'm not that strictly opposed to sending fields we don't use, but this one looked like you may be sending way too much data, so I was wondering if something else could be wrong.

But great, `_.pick` and `_.omit` should probably help :).

(`_.pick(modelObject, 'id', 'name', 'ext_management_system.id' ,...)` will return an object with just those properties you mentioned)

gildub commented:
@himdel, I reduced the sent data, thanks for underlining this. Also slight changes in the fields order, I provided screenshoots.

martinpovolny commented:
@gildub : ping?

gildub commented:
@martinpovolny, rebased and retested.

gildub commented:
@ZitaNemeckova, could you please review?

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,35 +1,78 @@
+          id: data.ext_management_system.id,
+          name: data.ext_management_system.name,
+        };
+        vm.networkRouterModel.cloud_tenant = { id: data.cloud_tenant.id,  name: data.cloud_tenant.name, };
```
Please have one param per line.

gildub replied:
Done.

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,35 +1,78 @@
+        if (data.extra_attributes.external_gateway_info && ! _.isEmpty(data.extra_attributes.external_gateway_info)) {
+          vm.networkRouterModel.external_gateway = true;
+          return getCloudSubnetsByRef(vm.networkRouterModel.extra_attributes.external_gateway_info.external_fixed_ips[0].subnet_id);
+        } else {
```
Code Climate's warning `Unnecessary 'else' after 'return'`. It means that anything in `else` should be instead out of that condition. Something like this:

```
if (...) {
   return ...;
}
vm.networkRouterModel.external_gateway = false;
```

gildub replied:
Done

ZitaNemeckova commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,35 +1,78 @@
-ManageIQ.angular.app.controller('networkRouterFormController', ['$http', '$scope', 'networkRouterFormId', 'miqService', function($http, $scope, networkRouterFormId, miqService) {
+ManageIQ.angular.app.controller('networkRouterFormController', ['networkRouterFormId', 'miqService', 'API', function(networkRouterFormId, miqService, API) {
```
Please add `$scope`. It's needed in on line 107 `$scope.angularForm.$setPristine(true);`

gildub replied:
Done

ZitaNemeckova commented:
@gildub You mentioned you reduce data for editing but there's still too much. 
```
href:http://localhost:3000/api/network_routers/10000000000005
name:cloudwest-router8
ems_id:10000000000014
admin_state_up:t
cloud_network_id:10000000000052
extra_attributes[distributed]:false
extra_attributes[high_availability]:false
extra_attributes[external_gateway_info][network_id]:6b3d0c3b-8b68-4c26-a493-5be44d160241
extra_attributes[external_gateway_info][enable_snat]:true
extra_attributes[external_gateway_info][external_fixed_ips][0][subnet_id]:1ca5cc3e-ffe1-44cf-94df-98a798489d06
extra_attributes[external_gateway_info][external_fixed_ips][0][ip_address]:10.9.60.144
id:10000000000005
cloud_subnets[0][href]:http://localhost:3000/api/cloud_subnets/10000000000055
cloud_subnets[0][id]:10000000000055
cloud_subnets[0][name]:cloudwest-subnet
cloud_subnets[0][ems_ref]:e725b7b2-6b8b-4b71-a868-5eb71f924834
cloud_subnets[0][ems_id]:10000000000014
cloud_subnets[0][availability_zone_id]:
cloud_subnets[0][cloud_network_id]:10000000000051
cloud_subnets[0][cidr]:192.168.202.0/24
cloud_subnets[0][status]:active
cloud_subnets[0][dhcp_enabled]:true
cloud_subnets[0][gateway]:192.168.202.1
cloud_subnets[0][network_protocol]:ipv4
cloud_subnets[0][cloud_tenant_id]:10000000000005
cloud_subnets[0][dns_nameservers][]:10.11.164.2
cloud_subnets[0][ipv6_router_advertisement_mode]:
cloud_subnets[0][ipv6_address_mode]:
cloud_subnets[0][extra_attributes][allocation_pools][0][start]:192.168.202.2
cloud_subnets[0][extra_attributes][allocation_pools][0][end]:192.168.202.254
cloud_subnets[0][extra_attributes][ip_version]:4
cloud_subnets[0][type]:ManageIQ::Providers::Openstack::NetworkManager::CloudSubnet
cloud_subnets[0][network_router_id]:10000000000005
cloud_subnets[0][network_group_id]:
cloud_subnets[0][parent_cloud_subnet_id]:
ext_management_system[id]:10000000000014
ext_management_system[name]:OpenStack Network Manager
cloud_tenant[id]:10000000000005
cloud_tenant[name]:cloudwest
external_gateway:true
cloud_subnet_id:10000000000056
```
But only this is used to update a router [here](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/controllers/network_router_controller.rb#L227):
```ruby
{:name=>"asimonel-routervvv",
 :ems_id=>"10000000000014",
 :cloud_subnet_id=>"10000000000056",
 :cloud_network_id=>"10000000000052",
 :external_gateway_info=>{:network_id=>"6b3d0c3b-8b68-4c26-a493-5be44d160241", 
                          :external_fixed_ips=>[{:subnet_id=>"1ca5cc3e-ffe1-44cf-94df-98a798489d06"}], 
                          :enable_snat=>"true"}}
```
So `cloud_subnets`, `ext_management_system`, `cloud_tenant`, `href`,  `admin_state_up` and part of `extra_attributes` are never used or passed anywhere.

ZitaNemeckova commented on app/views/network_router/_common_new_edit.haml:
```diff
@@ -7,25 +7,56 @@
+        .col-md-8{"ng-if" => "!vm.newRecord"}
+          %input.form-control{"type"         => "text",
+            "name"         => "ems_name",
+            "ng-maxlength" => 128,
```
Use `maxlength` as it's a static value.

gildub replied:
Done

ZitaNemeckova commented on app/views/network_router/_common_new_edit.haml:
```diff
@@ -7,25 +7,56 @@
+        .col-md-8{"ng-if" => "!vm.newRecord"}
+          %input.form-control{"type"         => "text",
+                              "name"         => "cloud_tenant_name",
+                              "ng-maxlength" => 128,
```
Use `maxlength` as it's a static value.

gildub replied:
Done

ZitaNemeckova commented on app/controllers/network_router_controller.rb:
```diff
@@ -124,8 +73,8 @@ def create
-      ems = ExtManagementSystem.find(options[:ems_id])
-      options.delete(:ems_id)
+
+      ems = ExtManagementSystem.find(params[:ext_management_system][:id])
```
I'm confused about `params[:ems_id]` and `params[:ext_management_system][:id]`. As it's 
 a same thing but you need one for Edit and another for Add. Please use one for both :)

gildub replied:
I'll make sure to use either one consistently, thanks

himdel replied:
So.. after https://github.com/ManageIQ/manageiq-ui-classic/pull/2422/commits/307e6c1394f576d680dd4013182917cbae3fad2a, this change is not needed at all here, right?

himdel commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -28,12 +34,59 @@ ManageIQ.angular.app.controller('networkRouterFormController', ['$http', '$scope
+      vm.networkRouterModel.ems_name = data.ext_management_system.name;
+      vm.networkRouterModel.cloud_tenant_name = data.cloud_tenant.name;
+      vm.networkRouterModel.extra_attributes = data.extra_attributes;
+    }).then(function() {
```
this line should probably not be here?

`data` is only defined in the first function, not the second (hence the eslint warning)

gildub replied:
Right! Done.

himdel commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -28,12 +34,59 @@ ManageIQ.angular.app.controller('networkRouterFormController', ['$http', '$scope
+        getSubnetByRef(ref);
+      }
+    }).then(function() {
+      getCloudNetworksByEms(vm.networkRouterModel.ems_id);
```
missing `return`

gildub replied:
Done

himdel commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -28,12 +34,59 @@ ManageIQ.angular.app.controller('networkRouterFormController', ['$http', '$scope
+    }).then(function() {
+      getCloudNetworksByEms(vm.networkRouterModel.ems_id);
+    }).then(function() {
+      getCloudSubnetsByNetworkID(vm.networkRouterModel.cloud_network_id);
```
missign `return`

gildub replied:
Done

gildub commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,55 +1,102 @@
+
+        vm.networkRouterModel.name = '';
+        vm.networkRouterModel.admin_state_up = true;
+//        vm.networkRouterModel.extra_attributes.external_gateway_info.enable_snat = true;
```
Sorry I'm working on it and I forgot to add [WIP] before I took off yesterday!

gildub commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,55 +1,102 @@
+          vm.networkRouterModel.external_gateway = true;
+          if (vm.networkRouterModel.extra_attributes.external_gateway_info.external_fixed_ips) {
+            var ref = vm.networkRouterModel.extra_attributes.external_gateway_info.external_fixed_ips[0].subnet_id;
+            getSubnetByRef(ref);
```
Yes

gildub commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,55 +1,102 @@
+
+    var getSubnetByRef = function(ref) {
+      if (ref) {
+        API.get("/api/cloud_subnets?expand=resources&attributes=name&filter[]=ems_ref=" + ref).then(function(data) {
```
Yes

gildub commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,55 +1,102 @@
+
+  vm.getCloudNetworksByEms = function(id) {
+    if (id) {
+      API.get("/api/cloud_networks?expand=resources&attributes=name,ems_ref&filter[]=external_facing=true&filter[]=ems_id=" + id).then(function(data) {
```
Yes

gildub commented on app/assets/javascripts/controllers/network_router/network_router_form_controller.js:
```diff
@@ -1,55 +1,102 @@
-  }
+  vm.getCloudSubnetsByNetworkID = function(id) {
+    if (id) {
+      API.get("/api/cloud_subnets?expand=resources&attributes=name,ems_ref&filter[]=cloud_network_id=" + id).then(function(data) {
```
yes

gildub commented:
@himdel, all green now.

gildub commented:
@martinpovolny, @ZitaNemeckova, sorry for stalling on this one, there is an independent issuehttps://github.com/ManageIQ/manageiq-ui-classic/issues/3918 I thought was related to this PR.

gildub commented:
@martinpovolny, @ZitaNemeckova, please review.

martinpovolny commented:
@ZitaNemeckova , @AparnaKarve : please, take care of this one. @himdel is busy with V2V. Thx!

AparnaKarve commented:
FYI, I am currently tied up with v2v blockers and deliverables for the upcoming sprint, and would not be able to review this for some time.
Thanks.

ZitaNemeckova commented on app/controllers/network_router_controller.rb:
```diff
@@ -462,14 +402,16 @@ def form_external_gateway(params)
         options[:external_gateway_info][:external_fixed_ips] = [{:subnet_id => subnet.ems_ref}]
       end
-      options[:external_gateway_info][:enable_snat] = switch_to_bool(params[:enable_snat])
+      if params[:extra_attributes][:external_gateway_info][:enable_snat]
```
Please use `params.fetch_path(:extra_attributes, :external_gateway_info, :enable_snat)` just in case `params` don't have `:extra_attributes` or `:external_gateway_info`.

gildub replied:
Done, thanks.

ZitaNemeckova commented on app/views/network_router/_common_new_edit.haml:
```diff
@@ -1,123 +1,140 @@
+    = _('Network Management Provider')
+
+  .form-horizontal
+    .form-group
```
This should be `.form-group{'ng-class' => '{"has-error": angularForm.ems_id.$invalid}'}` the selectpicker and Required text turn red.

gildub replied:
Good catch! Done, thanks.

ZitaNemeckova commented:
LGTM 👍 except two comments [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/2422#pullrequestreview-123943077) and [here](https://github.com/ManageIQ/manageiq-ui-classic/pull/2422#pullrequestreview-123882595).

gildub commented:
@ZitaNemeckova, thanks for taking over.
I've updated per your last comment.
I'm pinging you here because of the TZ difference ;)

ZitaNemeckova commented:
LGTM 👍

himdel commented on app/controllers/network_router_controller.rb:
```diff
@@ -451,14 +391,16 @@ def form_external_gateway(params)
   def form_params(params)
-    options = %i(name ems_id admin_state_up cloud_group_id cloud_subnet_id
-      cloud_network_id).each_with_object({}) do |param, opt|
+    options = %i(name admin_state_up ems_id cloud_group_id cloud_subnet_id
```
`ems_id`? Is this still being used? Because if so, you can't be using `ext_management_system.id` for the same purpose :)

gildub replied:
Not using `ems_id. I replaced it, thanks.

ZitaNemeckova commented on spec/controllers/network_router_controller_spec.rb:
```diff
@@ -123,7 +128,12 @@
+          :args        => [{
+            :name            => 'test',
+            :admin_state_up  => 'true',
+            :ems_id          => "@ems.id",
```
Test failure for Ruby is related to this line. Do you really want to have `"@ems.id"` or `@ems.id` as a string. If latter you should use `@ems.id.to_s`. Travis has to be green :)

himdel commented:
(restarting travis to make sure network_router_controller_spec is the only remaining spec failure)

gildub commented:
@himdel, good after restarting and fixing test format, thanks

ZitaNemeckova commented:
@gildub Sorry but I have to insist that Provider's id has to be send under same key for Add and Edit request. For Add you use `ems_id` and for Edit you use `ext_management_system[id]`. 

And I'm bit confused why when Adding Router you send `cloud_tenant[id]` and when Editing you send `cloud_tenant[name]`.

ZitaNemeckova commented on app/views/network_router/_common_new_edit.haml:
```diff
@@ -1,123 +1,140 @@
-                "ng-change"                   => "vm.filterCloudNetworkChanged(vm.networkRouterModel.cloud_network_id)",
-                "pf-select"                   => true,
-                "ng-selected"                 => "vm.networkRouterModel.cloud_network_id",
-                "ng-required"                 => "vm.networkRouterModel.enable_snat == true",
```
@gildub Did you remove `ng-required` on purpose?

Hyperkid123 commented:
The form works fine. I could not find anything that is not mentioned here. But i would appreciate some specs for the JavaScript side. Even though there is just a small amount of logic, i think it would be nice to cover it and make it a bit more future proof.

gildub commented:
@Hyperkid123, the driver here is to re-factor to add API get, the next step will be to replace with controller to component conversion (such as [1] for example) where the framework for testing is much better therefore I think it might not worth the effort.

[1] https://github.com/ManageIQ/manageiq-ui-classic/pull/2828

Hyperkid123 commented:
@gildub that sounds great

gildub commented:
@ZitaNemeckova, @Hyperkid123, then if everybody is happy any chance we can merge? Thanks

Hyperkid123 commented:
I think its fine. There some codeclimate issues that should be fixed though (the missing semicolons for example, not the similar code warnings etc.).

But i think it should be merged.
